[9593] Align IRequest/Request getHeader/setHeader types

twm opened the pull request:
Make the documentation match the implementation.

## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/9593#comment:1
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests. N/A

twm committed:
Align IRequest/Request getHeader/setHeader types

twm committed:
Add a newsfile

wiml commented on src/twisted/web/iweb.py:
```diff
@@ -53,18 +53,26 @@ def getHeader(key):
         Get an HTTP request header.
 
-        @type key: C{str}
+        @type key: C{bytes} or C{str}
```
The [Twisted coding standard page](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#docstrings) says that even references to Python's builtin types should use `L{}` instead of `C{}`.

twm replied:
:+1: Good point, I was just mindlessly copying the existing style.

wiml commented on src/twisted/web/http.py:
```diff
@@ -974,12 +974,13 @@ def getHeader(self, key):
         """
         Get an HTTP request header.
 
-        @type key: C{bytes}
```
Is it necessary to duplicate interfaces' docstrings on their implementing classes when they're identical? Current practice is mixed, and it seems redundant, since the generated documentation will come out the same either way (the implementing class's methods will be documented with text taken from the interface).

wiml replied:
(I should add that that comment isn't intended as an objection to this PR.)

twm replied:
I know that I've seen pydoctor inherit the docstrings before, so I removed them dupes from this PR, but it doesn't seem to be working properly in the case of `http.Request` or `server.Request`:

![image](https://user-images.githubusercontent.com/43662/55286043-31858f80-534b-11e9-8659-91130292c832.png)

![image](https://user-images.githubusercontent.com/43662/55286039-23377380-534b-11e9-8100-768a922f0ea9.png)

pydoctor is especially confused in the latter case &mdash; it lists the `Componentized` methods several times.

hawkowl replied:
(note: pydoctor's confusion there is fixed on current pydoctor, as seen on 18.9s docs)

twm replied:
Yes, the `Componentized` bit is fixed with pydoctor trunk.

It still doesn't inherit the documentation, though. `method`, `uri`, `path`, and `args` aren't shown, and the methods show up as undocumented:

![image](https://user-images.githubusercontent.com/43662/55293059-b18d1300-53a6-11e9-939d-73f0f168ac55.png)

twm replied:
Maybe this is happening because of the mix of inheritance and interface annotation here?

```
http.Request      implements    IConsumer
     |            defines       setResponseCode
     |                          setHeader
     |
server.Request    implements    IRequest
```

In any case, since `http.Request` doesn't implement `IRequest` it needs to document these methods and attributes on its own. (Indeed, there are other subclasses of `http.Request` that don't implement `IRequest` either — `ProxyRequest` and `ReverseProxyRequest`).

twm committed:
Merge remote-tracking branch 'origin/trunk' into 9593-http-request-header-types

twm committed:
Remove duplicate docs, per review

pydoctor knows how to take the docs from the interface, so they need not
be repeated in the implementation.

twm committed:
C{type} to L{type}, per review

twm committed:
Merge branch 'trunk' into 9593-http-request-header-types

twm committed:
Ooops, missed one

twm committed:
Fix reference to BytesIO

twm committed:
Fix _sanitizeLinearWhitespace epytext syntax

twm committed:
:'(

twm committed:
Restore http.Request docstrings

Because http.Request doesn't implement IRequest we can't rely on the
docstrings from the interface.

twm committed:
Fix linting issues

twm committed:
Update Hyper h2 and priority intersphinx URLs

glyph committed:
Merge branch 'trunk' into 9593-http-request-header-types

glyph committed:
Merge branch 'trunk' into 9593-http-request-header-types

glyph reviewed (approved):
My only quibble with this as-is is that it's such a big upgrade to the docs for this super core interface that I feel like it might deserve more than a `.misc` ;).  But by all means, let's land it soon!

twm committed:
Add a .doc newsfile

twm commented:
Thank you for the review, @glyph! I've added a .doc newsfile and will merge when the build goes green.

twm commented:
Now all the buildbots are failing like this:

```
Traceback (most recent call last):
  File "c:\buildslave\windows7-64-py2_7-coverage\twisted\build\py27-alldeps-withcov-windows\lib\site-packages\twisted\python\test\test_url.py", line 486, in test_parseEqualSignInParamValue
    self.assertEqual(u.asText(), 'http://localhost/?=x%3Dx%3Dx')
  File "c:\buildslave\windows7-64-py2_7-coverage\twisted\build\py27-alldeps-withcov-windows\lib\site-packages\twisted\trial\_synctest.py", line 432, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "c:\python27\Lib\unittest\case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "c:\python27\Lib\unittest\case.py", line 924, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
twisted.trial.unittest.FailTest: u'http://localhost/?=x=x=x' != u'http://localhost/?=x%3Dx%3Dx'
- http://localhost/?=x=x=x
?                     ^ ^
+ http://localhost/?=x%3Dx%3Dx
?                     ^^^ ^^^
```

The failure seems legitimate, so why weren't they before? I am so confused. -.-

twm commented:
![image](https://user-images.githubusercontent.com/43662/55774908-f4f11c80-5a4b-11e9-9679-26bd08e2e1dd.png)

New hyperlink?

twm commented:
Yes: https://twistedmatrix.com/trac/ticket/9619 :sob:

wiml committed:
Merge branch 'trunk' into 9593-http-request-header-types
